api/auth.proto

```diff
@@ -0,0 +1,33 @@
+syntax = "proto3";
+
+package envoy.api.v2;
+
+message AuthAction {
+
+  enum VerifyType {
+    option allow_alias = true;
+
+    DEFAULT = 0;
+    WHITE_LIST = 0;
+    BLACK_LIST = 1;
+  }
+
+  // Check peer identity using X.509 certificate.
+  message X509 {
+    // SHA256 hash of peer certificate.
+    string sha256_hash = 1;
+    // Subject of peer certificate.
+    string subject = 2;
+    // Subject alt name of peer certificate.
+    string subject_alt_name = 3;
+  }
+
+  message X509Verify {
+    // Do white-list or black-list verify.
+    VerifyType verify_type = 1;
+    // List of X509 to check against.
+    repeated X509 x509s = 2;
+  }
+
+  X509Verify x509_verify = 1;
```

Comment:
I think my previous comment was lost. Ultimately, we need to potentially add other things beyond just X509 to auth on. For example, once we have IP tagging, we might only want to allow certain tags which will be inside a header name.

Can we future proof this by making the top level object something like:
```
repeated VerifyRule rules = 1;
```

And then having VerifyRule be a oneof that can contain X509Verify, but also HeaderVerify, etc. in the future?

Reply:
i did the "rule as list" part, but there are two approach regarding verify logic:
1. control all rule by a single allow/deny config; 
2. each rule have its own allow/deny config
i was taking the first route here.